src/utilcode/longfilepathwrappers.cpp

```diff
@@ -0,0 +1,1307 @@
+    {
+        //Change the behaviour in Redstone to retry
+        COUNT_T size = MAX_LONGPATH;
+
+        ret = GetCurrentDirectoryW(
+            size, 
+            lpBuffer.OpenUnicodeBuffer(size - 1)
+            );
+
+        lastError = GetLastError();
+        lpBuffer.CloseBuffer(ret);
+    }
+    EX_CATCH_HRESULT(hr);
+
+    if (hr != S_OK)
+    {
+        SetLastError(hr);
+    }
+    else if (ret == 0)
+    {
+        SetLastError(lastError);
+    }
+
+    return ret;
+}
+
+DWORD WINAPI GetEnvironmentVariableWrapper(
+    _In_opt_  LPCTSTR lpName,
+    _Out_opt_ SString&  lpBuffer
+    )
+{
+    HRESULT hr = S_OK;
+    DWORD ret = 0;
+    DWORD lastError;
+    EX_TRY
+    {
+        
+        COUNT_T size = lpBuffer.GetUnicodeAllocation() + 1;
+
+        ret = GetEnvironmentVariableW(
+            lpName, 
+            lpBuffer.OpenUnicodeBuffer(size - 1),
+            size
+            );
+
+        if (ret > size)
+        {
+            lpBuffer.CloseBuffer();
+            ret = GetEnvironmentVariableW(
+                lpName, 
+                lpBuffer.OpenUnicodeBuffer(ret - 1), 
+                ret);
+        }
+
+        lastError = GetLastError();
+        lpBuffer.CloseBuffer(ret);
+    }
+    EX_CATCH_HRESULT(hr);
+
+    if (hr != S_OK)
+    {
+        SetLastError(hr);
+    }
+    else if (ret == 0)
+    {
+        SetLastError(lastError);
+    }
+
+    return ret;
+}
+
+
+#ifndef FEATURE_PAL
+
+BOOL
+CreateHardLinkWrapper(
+        _In_       LPCWSTR lpFileName,
+        _In_       LPCWSTR lpExistingFileName,
+        _Reserved_ LPSECURITY_ATTRIBUTES lpSecurityAttributes
+        )
+{
+    HRESULT hr = S_OK;
+    BOOL ret   = FALSE;
+    DWORD lastError;
+    EX_TRY
+    {
+        LongPathString Existingpath(LongPathString::Literal, lpExistingFileName);
+        LongPathString FileName(LongPathString::Literal, lpFileName);
+
+        if (SUCCEEDED(LongFile::NormalizePath(Existingpath)) && SUCCEEDED(LongFile::NormalizePath(FileName)))
+        {
+            ret = CreateHardLinkW(
+                    Existingpath.GetUnicode(),
+                    FileName.GetUnicode(),
+                    lpSecurityAttributes
+                    );
+        }
+        
+        lastError = GetLastError();
+    }
+    EX_CATCH_HRESULT(hr);
+
+    if (hr != S_OK )
+    {
+        SetLastError(hr);
+    }
+    else if(ret == FALSE)
+    {
+        SetLastError(lastError);
+    }
+
+    return ret;
+
+}
+
+BOOLEAN
+CreateSymbolicLinkWrapper(
+        _In_ LPCWSTR lpSymlinkFileName,
+        _In_ LPCWSTR lpTargetFileName,
+        _In_ DWORD dwFlags
+        )
+{
+    HRESULT hr  = S_OK;
+    BOOLEAN ret = FALSE;
+    DWORD lastError;
+    EX_TRY
+    {
+        LongPathString TargetFileName(LongPathString::Literal, lpTargetFileName);
+        LongPathString linkFileName(LongPathString::Literal, lpSymlinkFileName);
+
+        if (SUCCEEDED(LongFile::NormalizePath(TargetFileName)) && SUCCEEDED(LongFile::NormalizePath(linkFileName)))
```

Comment:
Shouldn't normalize TargetFileName. When a relative TargetFileName is used, it is supposed to be relative to linkFileName, not relative to current directory. Even if you fix that, it's still incorrect to convert relative path to absolute path before calling CreateSymbolicLinkW. A link can be moved after it is created, and its target should move accordingly. In short, a relative symbolic link is semantically different from an absolute one.

We don't seem to use CreateSymbolicLink any way, so we don't really need a wrapper for it.